c/polyjuice.h

```diff
@@ -1366,27 +1373,36 @@ int run_polyjuice() {
   uint8_t evm_memory[MAX_EVM_MEMORY_SIZE];
   init_evm_memory(evm_memory, MAX_EVM_MEMORY_SIZE);
 
+  /* init EVM execution result */
   struct evmc_result res;
   memset(&res, 0, sizeof(evmc_result));
-  res.status_code = EVMC_FAILURE; // Generic execution failure
+  res.status_code = EVMC_FAILURE;      // Generic execution failure
+  debug_print_int("[run_polyjuice] initial gas limit", msg.gas);
+  int64_t initial_gas = msg.gas;
+  msg.gas -= MIN_TRANSACTION_GAS;      // subtract IntrinsicGas
 
   int ret_handle_message = handle_message(&context, UINT32_MAX, UINT32_MAX, NULL, &msg, &res);
-  uint64_t gas_used = (uint64_t)(msg.gas - res.gas_left);
-
   // debug_print evmc_result.output_data if the execution failed
   if (res.status_code != 0) {
     debug_print_int("evmc_result.output_size", res.output_size);
     // The output contains data coming from REVERT opcode
     debug_print_data("evmc_result.output_data:", res.output_data,
                      res.output_size > 100 ? 100 : res.output_size);
   }
-  
+
+  debug_print_int("[run_polyjuice] gas left", res.gas_left);
+  uint64_t gas_used =
+      (uint64_t)(res.gas_left <= 0 ? initial_gas : initial_gas - res.gas_left);
```

Comment:
Shouldn't we report error when `res.gas_left < 0`?

Reply:
https://github.com/nervosnetwork/godwoken-polyjuice/blob/159378db76979bfd92f11cfc33bbd6e45d01941d/c/polyjuice.h#L1024-L1028

The transaction will turn out to be a failed transaction with EVMC_OUT_OF_GAS exit_code.
And then, Godwoken will [handle tx fee here](https://github.com/nervosnetwork/godwoken/blob/76b43a1762f5cb703aeeafc4a31ec67e5d543bb5/crates/generator/src/generator.rs#L677-L711).

Note: In the latest version of Godwoken, failed transactions are packed into Layer2 block.